File: php/src/Mail.php
<?php
declare(strict_types=1);

final class Mailer
{
    public static function configured(): bool
    {
        return self::resendKey() !== '' || self::smtpConfigured();
    }

    /** @return array{configured:bool,provider:string,resend:bool,smtp:bool,mail_from:string,smtp_host:string,smtp_port:int} */
    public static function status(): array
    {
        $resend = self::resendKey() !== '';
        $smtp = self::smtpConfigured();
        $force = strtolower(trim(Env::get('MAIL_PROVIDER')));
        $provider = 'none';
        if ($force === 'smtp' && $smtp) {
            $provider = 'smtp';
        } elseif ($force === 'resend' && $resend) {
            $provider = 'resend';
        } elseif ($resend && $smtp) {
            $provider = 'resend_then_smtp';
        } elseif ($resend) {
            $provider = 'resend';
        } elseif ($smtp) {
            $provider = 'smtp';
        }
        return [
            'configured' => self::configured(),
            'provider' => $provider,
            'resend' => $resend,
            'smtp' => $smtp,
            'mail_from' => self::mailFrom(),
            'smtp_host' => $smtp ? trim(Env::get('SMTP_HOST')) : '',
            'smtp_port' => $smtp ? self::smtpPort() : 0,
        ];
    }

    public static function send(string $to, string $subject, string $body, ?string $fromName = null, array $attachments = []): array
    {
        $to = trim($to);
        if ($to === '' || !str_contains($to, '@')) {
            throw new RuntimeException('Client has no email address');
        }
        $mailFrom = self::mailFrom();
        if ($mailFrom === '') {
            throw new RuntimeException('MAIL_FROM is not set');
        }
        $display = trim($fromName ?: Env::get('MAIL_FROM_NAME', 'InPmnt'));
        $fromHeader = self::fromHeader($display, $mailFrom);

        $force = strtolower(trim(Env::get('MAIL_PROVIDER')));
        $resend = self::resendKey();
        $smtpOk = self::smtpConfigured();
        $tryResend = $resend !== '' && $force !== 'smtp';
        $trySmtp = $smtpOk && ($force === 'smtp' || !$tryResend);

        $errors = [];
        if ($tryResend) {
            try {
                self::resend($resend, $fromHeader, $to, $subject, $body, $attachments);
                return ['provider' => 'resend', 'id' => null];
            } catch (Throwable $e) {
                $errors[] = $e->getMessage();
                if (!$smtpOk || $force === 'resend') {
                    throw $e;
                }
                $trySmtp = true;
            }
        }
        if ($trySmtp) {
            $host = trim(Env::get('SMTP_HOST'));
            try {
                self::smtpSocket($host, $fromHeader, $mailFrom, $to, $subject, $body, $attachments);
                return ['provider' => 'smtp', 'id' => null];
            } catch (Throwable $e) {
                if ($errors) {
                    throw new RuntimeException($errors[0] . '; SMTP fallback: ' . $e->getMessage());
                }
                throw $e;
            }
        }
        throw new RuntimeException(
            'Email is not configured. Set RESEND_API_KEY or SMTP_HOST/SMTP_USER/MAIL_FROM in .env'
        );
    }

    private static function mailFrom(): string
    {
        return trim(Env::get('MAIL_FROM') ?: Env::get('SMTP_USER'));
    }

    private static function resendKey(): string
    {
        $key = trim(Env::get('RESEND_API_KEY'));
        if ($key === '' || str_starts_with($key, 're_...')) {
            return '';
        }
        $lower = strtolower($key);
        if ($lower === 're_your_api_key' || str_contains($lower, 'changeme')) {
            return '';
        }
        return $key;
    }

    private static function smtpConfigured(): bool
    {
        return trim(Env::get('SMTP_HOST')) !== ''
            && trim(Env::get('SMTP_USER')) !== ''
            && self::mailFrom() !== '';
    }

    private static function smtpPort(): int
    {
        return (int) (Env::get('SMTP_PORT', '587') ?: '587');
    }

    private static function fromHeader(string $display, string $mailFrom): string
    {
        $display = str_replace(["\r", "\n"], '', $display);
        if ($display === '') {
            return $mailFrom;
        }
        if (preg_match('/[^\x20-\x7E]/', $display) || strpbrk($display, '",<>') !== false) {
            $encoded = '=?UTF-8?B?' . base64_encode($display) . '?=';
            return "{$encoded} <{$mailFrom}>";
        }
        return "{$display} <{$mailFrom}>";
    }

    private static function headerSafe(string $value): string
    {
        $value = str_replace(["\r", "\n"], '', $value);
        if ($value !== '' && preg_match('/[^\x20-\x7E]/', $value)) {
            return '=?UTF-8?B?' . base64_encode($value) . '?=';
        }
        return $value;
    }

    private static function resend(string $apiKey, string $from, string $to, string $subject, string $body, array $attachments = []): void
    {
        if (!function_exists('curl_init')) {
            throw new RuntimeException('Resend error: PHP curl extension is not enabled');
        }
        $payloadArr = [
            'from' => $from,
            'to' => [$to],
            'subject' => $subject !== '' ? $subject : '(no subject)',
            'text' => $body,
        ];
        if ($attachments) {
            $payloadArr['attachments'] = [];
            foreach ($attachments as $att) {
                $payloadArr['attachments'][] = [
                    'filename' => (string) ($att['filename'] ?? 'invoice.pdf'),
                    'content' => base64_encode((string) ($att['content'] ?? '')),
                ];
            }
        }
        $payload = json_encode($payloadArr, JSON_UNESCAPED_SLASHES);
        $ch = curl_init('https://api.resend.com/emails');
        curl_setopt_array($ch, [
            CURLOPT_POST => true,
            CURLOPT_HTTPHEADER => [
                'Authorization: Bearer ' . $apiKey,
                'Content-Type: application/json',
            ],
            CURLOPT_POSTFIELDS => $payload,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => 30,
        ]);
        $raw = curl_exec($ch);
        $code = (int) curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $err = curl_error($ch);
        curl_close($ch);
        if ($raw === false) {
            throw new RuntimeException('Resend error: ' . $err);
        }
        if ($code >= 400) {
            throw new RuntimeException("Resend error {$code}: {$raw}");
        }
    }

    private static function smtpSocket(
        string $host,
        string $fromHeader,
        string $mailFrom,
        string $to,
        string $subject,
        string $body,
        array $attachments = []
    ): void {
        $port = self::smtpPort();
        $user = trim(Env::get('SMTP_USER'));
        $password = Env::get('SMTP_PASSWORD');
        $useSsl = Env::truthy('SMTP_SSL') || $port === 465;
        $wantStartTls = !$useSsl && Env::get('SMTP_STARTTLS', '1') !== '0';

        $err = '';
        $fp = self::smtpConnect($host, $port, $useSsl, true, $err);
        if ($fp === false) {
            $fp = self::smtpConnect($host, $port, $useSsl, false, $err);
        }
        if ($fp === false) {
            throw new RuntimeException("SMTP connect failed: {$host}:{$port} {$err}");
        }
        stream_set_timeout($fp, 30);

        try {
            $read = static function () use ($fp): string {
                $line = '';
                while (!feof($fp)) {
                    $chunk = fgets($fp, 515);
                    if ($chunk === false) {
                        break;
                    }
                    $line .= $chunk;
                    if (isset($chunk[3]) && $chunk[3] === ' ') {
                        break;
                    }
                }
                return $line;
            };
            $cmd = static function (string $c) use ($fp, $read): string {
                fwrite($fp, $c . "\r\n");
                return $read();
            };
            $expect = static function (string $resp, string $okPrefix, string $what): string {
                if (!str_starts_with(ltrim($resp), $okPrefix)) {
                    $trim = trim(str_replace(["\r", "\n"], ' ', $resp));
                    throw new RuntimeException("SMTP {$what} failed: {$trim}");
                }
                return $resp;
            };

            $greeting = $read();
            $expect($greeting, '220', 'connect');
            $ehlo = $cmd('EHLO inpmnt.local');
            if (!str_starts_with(ltrim($ehlo), '250')) {
                $ehlo = $cmd('HELO inpmnt.local');
                $expect($ehlo, '250', 'HELO');
            }
            if ($wantStartTls) {
                $tls = $cmd('STARTTLS');
                $expect($tls, '220', 'STARTTLS');
                $cryptoOk = @stream_socket_enable_crypto($fp, true, STREAM_CRYPTO_METHOD_TLS_CLIENT);
                if ($cryptoOk !== true) {
                    throw new RuntimeException('SMTP STARTTLS handshake failed');
                }
                $cmd('EHLO inpmnt.local');
            }
            if ($user !== '') {
                self::smtpAuth($cmd, $expect, $user, $password);
            }
            $expect($cmd('MAIL FROM:<' . $mailFrom . '>'), '250', 'MAIL FROM');
            $expect($cmd('RCPT TO:<' . $to . '>'), '25', 'RCPT TO');
            $expect($cmd('DATA'), '354', 'DATA');
            $safeSubject = self::headerSafe($subject !== '' ? $subject : '(no subject)');
            $payload = self::rfc822($fromHeader, $mailFrom, $to, $safeSubject, $body, $attachments);
            $expect($cmd($payload), '250', 'message');
            $cmd('QUIT');
        } finally {
            fclose($fp);
        }
    }

    /** @param list<array{filename?:string,content?:string,mime?:string}> $attachments */
    private static function rfc822(
        string $fromHeader,
        string $mailFrom,
        string $to,
        string $safeSubject,
        string $body,
        array $attachments
    ): string {
        $dot = static function (string $s): string {
            $s = str_replace(["\r\n", "\r"], "\n", $s);
            $s = preg_replace('/^\./m', '..', $s) ?? $s;
            return str_replace("\n", "\r\n", $s);
        };
        $headers = 'Subject: ' . $safeSubject . "\r\n"
            . 'From: ' . $fromHeader . "\r\n"
            . 'To: ' . $to . "\r\n"
            . 'Reply-To: ' . $mailFrom . "\r\n"
            . "MIME-Version: 1.0\r\n";
        if (!$attachments) {
            return $dot(
                $headers
                . "Content-Type: text/plain; charset=UTF-8\r\n"
                . "Content-Transfer-Encoding: 8bit\r\n\r\n"
                . $body
            ) . "\r\n.";
        }
        $bound = '----=_InPmnt_' . bin2hex(random_bytes(8));
        $parts = $headers
            . 'Content-Type: multipart/mixed; boundary="' . $bound . "\"\r\n\r\n"
            . '--' . $bound . "\r\n"
            . "Content-Type: text/plain; charset=UTF-8\r\n"
            . "Content-Transfer-Encoding: 8bit\r\n\r\n"
            . $body . "\r\n";
        foreach ($attachments as $att) {
            $name = preg_replace('/[^A-Za-z0-9._-]+/', '_', (string) ($att['filename'] ?? 'invoice.pdf')) ?: 'invoice.pdf';
            $mime = (string) ($att['mime'] ?? 'application/pdf');
            $b64 = chunk_split(base64_encode((string) ($att['content'] ?? '')), 76, "\r\n");
            $parts .= '--' . $bound . "\r\n"
                . 'Content-Type: ' . $mime . '; name="' . $name . "\"\r\n"
                . "Content-Transfer-Encoding: base64\r\n"
                . 'Content-Disposition: attachment; filename="' . $name . "\"\r\n\r\n"
                . $b64;
        }
        $parts .= '--' . $bound . "--\r\n";
        return $dot($parts) . '.';
    }

    /** @param callable(string):string $cmd */
    private static function smtpAuth(callable $cmd, callable $expect, string $user, string $password): void
    {
        $login = $cmd('AUTH LOGIN');
        if (str_starts_with(ltrim($login), '334')) {
            $cmd(base64_encode($user));
            $resp = $cmd(base64_encode($password));
            if (str_starts_with(ltrim($resp), '235')) {
                return;
            }
            $plain = $cmd('AUTH PLAIN ' . base64_encode("\0{$user}\0{$password}"));
            if (str_starts_with(ltrim($plain), '235')) {
                return;
            }
            throw new RuntimeException('SMTP login failed: ' . trim(str_replace(["\r", "\n"], ' ', $resp)));
        }
        $plain = $cmd('AUTH PLAIN ' . base64_encode("\0{$user}\0{$password}"));
        $expect($plain, '235', 'AUTH PLAIN');
    }

    /** @param-out string $err @return resource|false */
    private static function smtpConnect(string $host, int $port, bool $ssl, bool $verify, string &$err)
    {
        $remote = ($ssl ? 'ssl://' : 'tcp://') . $host . ':' . $port;
        $ctx = stream_context_create([
            'ssl' => [
                'verify_peer' => $verify,
                'verify_peer_name' => $verify,
                'allow_self_signed' => !$verify,
                'SNI_enabled' => true,
                'peer_name' => $host,
            ],
        ]);
        $errno = 0;
        $errstr = '';
        $fp = @stream_socket_client(
            $remote,
            $errno,
            $errstr,
            30,
            STREAM_CLIENT_CONNECT,
            $ctx
        );
        if ($fp === false) {
            $err = trim($errstr . ($errno ? " ({$errno})" : ''));
        }
        return $fp;
    }
}

File: app/mail.py
"""Outbound email for reminders (Resend API or SMTP)."""
from __future__ import annotations

import os
import smtplib
import ssl
from email.message import EmailMessage
from email.utils import formataddr
from typing import Any, Sequence


_PLACEHOLDER_RESEND = {"re_...", "re_your_api_key", "re_changeme"}


def _clean(value: str | None) -> str:
    return (value or "").strip().strip('"').strip("'")


def _resend_key() -> str:
    key = _clean(os.environ.get("RESEND_API_KEY"))
    if not key:
        return ""
    lower = key.lower()
    if lower in _PLACEHOLDER_RESEND or lower.startswith("re_...") or "changeme" in lower:
        return ""
    return key


def _mail_from() -> str:
    return _clean(os.environ.get("MAIL_FROM") or os.environ.get("SMTP_USER"))


def smtp_configured() -> bool:
    # Hostinger SMTP rejects empty passwords after a long hang; treat as not configured.
    return bool(
        _clean(os.environ.get("SMTP_HOST"))
        and _clean(os.environ.get("SMTP_USER"))
        and _clean(os.environ.get("SMTP_PASSWORD"))
        and _mail_from()
    )


def mail_configured() -> bool:
    return bool(_resend_key() or smtp_configured())


def mail_status() -> dict[str, Any]:
    resend = bool(_resend_key())
    smtp = smtp_configured()
    force = _clean(os.environ.get("MAIL_PROVIDER")).lower()
    provider = "none"
    if force == "smtp" and smtp:
        provider = "smtp"
    elif force == "resend" and resend:
        provider = "resend"
    elif resend and smtp:
        provider = "resend_then_smtp"
    elif resend:
        provider = "resend"
    elif smtp:
        provider = "smtp"
    port = int(_clean(os.environ.get("SMTP_PORT")) or "587") if smtp else 0
    return {
        "configured": mail_configured(),
        "provider": provider,
        "resend": resend,
        "smtp": smtp,
        "mail_from": _mail_from() if mail_configured() else "",
        "smtp_host": _clean(os.environ.get("SMTP_HOST")) if smtp else "",
        "smtp_port": port,
    }


def send_email(
    *,
    to: str,
    subject: str,
    body: str,
    from_name: str | None = None,
    attachments: Sequence[dict[str, Any]] | None = None,
) -> dict[str, Any]:
    """Send a plain-text email, optionally with attachments. Raises RuntimeError on failure."""
    to = (to or "").strip()
    if not to or "@" not in to:
        raise RuntimeError("Client has no email address")

    mail_from = _mail_from()
    if not mail_from:
        raise RuntimeError("MAIL_FROM is not set")

    display = (from_name or os.environ.get("MAIL_FROM_NAME") or "InPmnt").strip()
    from_header = formataddr((display, mail_from)) if display else mail_from
    atts = list(attachments or [])

    force = _clean(os.environ.get("MAIL_PROVIDER")).lower()
    resend_key = _resend_key()
    smtp_ok = smtp_configured()
    try_resend = bool(resend_key) and force != "smtp"
    try_smtp = smtp_ok and (force == "smtp" or not try_resend)

    errors: list[str] = []
    if try_resend:
        try:
            return _send_resend(resend_key, from_header, to, subject, body, atts)
        except Exception as exc:  # noqa: BLE001
            errors.append(str(exc))
            if not smtp_ok or force == "resend":
                if isinstance(exc, RuntimeError):
                    raise
                raise RuntimeError(str(exc)) from exc
            try_smtp = True

    if try_smtp:
        host = _clean(os.environ.get("SMTP_HOST"))
        try:
            return _send_smtp(host, from_header, mail_from, to, subject, body, atts)
        except Exception as exc:  # noqa: BLE001
            detail = str(exc)
            if errors:
                raise RuntimeError(f"{errors[0]}; SMTP fallback: {detail}") from exc
            raise RuntimeError(detail) from exc

    raise RuntimeError(
        "Email is not configured. Set RESEND_API_KEY or SMTP_HOST/SMTP_USER/MAIL_FROM in .env"
    )


def _attachment_bytes(att: dict[str, Any]) -> bytes:
    raw = att.get("content") or b""
    if isinstance(raw, str):
        return raw.encode("latin-1")
    return bytes(raw)


def _send_resend(
    api_key: str,
    from_header: str,
    to: str,
    subject: str,
    body: str,
    attachments: Sequence[dict[str, Any]] | None = None,
) -> dict[str, Any]:
    import base64
    import json
    import urllib.error
    import urllib.request

    payload_obj: dict[str, Any] = {
        "from": from_header,
        "to": [to],
        "subject": subject or "(no subject)",
        "text": body or "",
    }
    if attachments:
        payload_obj["attachments"] = [
            {
                "filename": str(att.get("filename") or "invoice.pdf"),
                "content": base64.b64encode(_attachment_bytes(att)).decode("ascii"),
            }
            for att in attachments
        ]
    payload = json.dumps(payload_obj).encode("utf-8")
    req = urllib.request.Request(
        "https://api.resend.com/emails",
        data=payload,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            raw = resp.read().decode("utf-8")
            data = json.loads(raw) if raw else {}
            return {"provider": "resend", "id": data.get("id")}
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"Resend error {exc.code}: {detail}") from exc
    except urllib.error.URLError as exc:
        raise RuntimeError(f"Resend error: {exc.reason}") from exc


def _send_smtp(
    host: str,
    from_header: str,
    mail_from: str,
    to: str,
    subject: str,
    body: str,
    attachments: Sequence[dict[str, Any]] | None = None,
) -> dict[str, Any]:
    port = int(_clean(os.environ.get("SMTP_PORT")) or "587")
    user = _clean(os.environ.get("SMTP_USER"))
    password = os.environ.get("SMTP_PASSWORD") or ""
    if password.startswith('"') and password.endswith('"') and len(password) >= 2:
        password = password[1:-1]
    use_ssl = _clean(os.environ.get("SMTP_SSL")).lower() in ("1", "true", "yes", "on")

    msg = EmailMessage()
    msg["Subject"] = subject or "(no subject)"
    msg["From"] = from_header
    msg["To"] = to
    msg["Reply-To"] = mail_from
    msg.set_content(body or "")
    for att in attachments or []:
        mime = str(att.get("mime") or "application/pdf")
        main, _, sub = mime.partition("/")
        msg.add_attachment(
            _attachment_bytes(att),
            maintype=main or "application",
            subtype=sub or "pdf",
            filename=str(att.get("filename") or "invoice.pdf"),
        )

    context = ssl.create_default_context()
    if use_ssl or port == 465:
        with smtplib.SMTP_SSL(host, port, context=context, timeout=30) as smtp:
            if user:
                smtp.login(user, password)
            smtp.send_message(msg, from_addr=mail_from, to_addrs=[to])
    else:
        with smtplib.SMTP(host, port, timeout=30) as smtp:
            smtp.ehlo()
            if _clean(os.environ.get("SMTP_STARTTLS") or "1").lower() not in ("0", "false", "no"):
                smtp.starttls(context=context)
                smtp.ehlo()
            if user:
                smtp.login(user, password)
            smtp.send_message(msg, from_addr=mail_from, to_addrs=[to])
    return {"provider": "smtp", "id": None}
